README.md
# Holidays

[![Continuous Integration](https://github.com/mll-lab/holidays/workflows/Continuous%20Integration/badge.svg)](https://github.com/mll-lab/holidays/actions)
[![Code Coverage](https://codecov.io/gh/mll-lab/holidays/branch/master/graph/badge.svg)](https://codecov.io/gh/mll-lab/holidays)
[![StyleCI](https://github.styleci.io/repos/414166520/shield?branch=master)](https://github.styleci.io/repos/414166520)

[![Latest Stable Version](https://poser.pugx.org/mll-lab/holidays/v/stable)](https://packagist.org/packages/mll-lab/holidays)
[![Total Downloads](https://poser.pugx.org/mll-lab/holidays/downloads)](https://packagist.org/packages/mll-lab/holidays)

## Installation

Install through composer

```sh
composer require mll-lab/holidays
```

## Usage

```php
use MLL\Holidays\BavarianHolidays;

// Call static methods on BavarianHolidays
```

### Custom Holidays

You can add custom holidays by registering a method that returns a map of holidays for a given year.
Set this up in a central place that always runs before your application, e.g. a bootstrap method.

```php
use MLL\Holidays\BavarianHolidays;

BavarianHolidays::$loadUserDefinedHolidays = static function (int $year): array {
    switch ($year) {
        case 2019:
            return ['22.03' => 'Day of the Tentacle'];
        default:
            return [];
    }
};
```

Custom holidays have precedence over the holidays inherent to this library.

## Changelog

See [`CHANGELOG.md`](CHANGELOG.md).

## Contributing

See [`CONTRIBUTING.md`](.github/CONTRIBUTING.md).

## License

This package is licensed using the MIT License.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

See [GitHub releases](https://github.com/mll-lab/holidays/releases).

## Unreleased

## v3.3.0

### Added

- Support `thecodingmachine/safe:^2`

## v3.2.0

### Added

- Add class `DateModification`

## v3.1.0

### Added

- Add method `BavarianHolidays::subBusinessDays()`

## v3.0.0

### Changed

- Clarify distinction between weekend and holidays and rename methods accordingly

## v2.1.0

### Changed

- Downgrade the required PHP version to `^7.2`

## v2.0.0

### Added

- Add `BavarianHolidays::$loadUserDefinedHolidays` for custom holidays

### Changed

- Rename `Holidays` to `BavarianHolidays` and generalize method names

## v1.0.0

### Added

- Class `Holidays` to ease dealing with working days at MLL


tests/Unit/ExampleTest.php
<?php

declare(strict_types=1);

namespace MLL\Holidays\Tests\Unit;

use MLL\Holidays\Example;
use PHPUnit\Framework;

class ExampleTest extends Framework\TestCase
{
    public function testGreetIncludesName(): void
    {
        $name = 'mll-lab';
        $example = new Example($name);

        self::assertStringContainsString($name, $example->greet());
    }
}


tests/Unit/HolidaysTest.php
<?php

declare(strict_types=1);

namespace MLL\Holidays\Tests\Unit;

use Carbon\Carbon;
use MLL\Holidays\Holidays;
use PHPUnit\Framework\TestCase;

final class HolidaysTest extends TestCase
{
    public function testName(): void
    {
        self::assertNull(Holidays::name(self::workingDayWednesday()));
        self::assertSame(Holidays::SAMSTAG, Holidays::name(self::saturday()));
        self::assertSame(Holidays::SONNTAG, Holidays::name(self::sunday()));
        self::assertSame(Holidays::KARFREITAG, Holidays::name(self::karfreitag2019()));
        self::assertSame(Holidays::OSTERSONNTAG, Holidays::name(self::easterSunday2019()));
    }

    /**
     * @dataProvider workingDays
     */
    public function testWorkingDays(Carbon $workingDay): void
    {
        self::assertTrue(Holidays::isMLLWorkingDay($workingDay));
        self::assertFalse(Holidays::isMLLHoliday($workingDay));
    }

    /**
     * @dataProvider holidays
     */
    public function testHolidays(Carbon $holiday): void
    {
        self::assertFalse(Holidays::isMLLWorkingDay($holiday));
        self::assertTrue(Holidays::isMLLHoliday($holiday));
    }

    public function testAddMllWorkingDays(): void
    {
        $saturday = self::saturday();
        $mondayAfter = self::saturday()->addDays(2);

        self::assertTrue(
            Holidays::addMLLWorkingDays($saturday, 1)
                ->isSameDay($mondayAfter),
            'Skips over sunday'
        );
        self::assertTrue(
            $saturday->isSameDay(self::saturday()),
            'Should not mutate the original date'
        );
    }

    protected static function workingDayWednesday(): Carbon
    {
        return Carbon::createStrict(2019, 10, 30);
    }

    protected static function karfreitag2019(): Carbon
    {
        return Carbon::createStrict(2019, 4, 19);
    }

    protected static function easterSunday2019(): Carbon
    {
        return Carbon::createStrict(2019, 4, 21);
    }

    protected static function saturday(): Carbon
    {
        return Carbon::createStrict(2019, 11, 2);
    }

    protected static function sunday(): Carbon
    {
        return Carbon::createStrict(2019, 11, 3);
    }

    /**
     * @return iterable<int, array{Carbon}>
     */
    public static function workingDays(): iterable
    {
        yield [self::workingDayWednesday()];
    }

    /**
     * @return iterable<int, array{Carbon}>
     */
    public static function holidays(): iterable
    {
        yield [self::karfreitag2019()];
        yield [self::easterSunday2019()];
        yield [self::saturday()];
        yield [self::sunday()];
    }
}


tests/Unit/BavarianHolidaysTest.php
<?php

declare(strict_types=1);

namespace MLL\Holidays\Tests\Unit;

use Carbon\Carbon;
use MLL\Holidays\BavarianHolidays;
use PHPUnit\Framework\TestCase;

final class BavarianHolidaysTest extends TestCase
{
    public function testNameHoliday(): void
    {
        self::assertNull(BavarianHolidays::nameHoliday(self::businessDayWednesday()));
        self::assertSame(BavarianHolidays::KARFREITAG, BavarianHolidays::nameHoliday(self::karfreitag2019()));
        self::assertSame(BavarianHolidays::OSTERSONNTAG, BavarianHolidays::nameHoliday(self::easterSunday2019()));
    }

    /**
     * @dataProvider businessDays
     */
    public function testBusinessDays(Carbon $businessDay): void
    {
        self::assertTrue(BavarianHolidays::isBusinessDay($businessDay));
        self::assertFalse(BavarianHolidays::isHoliday($businessDay));
    }

    /**
     * @dataProvider holidays
     */
    public function testHolidays(Carbon $holiday): void
    {
        self::assertFalse(BavarianHolidays::isBusinessDay($holiday));
        self::assertTrue(BavarianHolidays::isHoliday($holiday));
    }

    /**
     * @dataProvider weekend
     */
    public function testWeekend(Carbon $weekend): void
    {
        self::assertFalse(BavarianHolidays::isBusinessDay($weekend));
        self::assertFalse(BavarianHolidays::isHoliday($weekend));
    }

    public function testAddMllBusinessDays(): void
    {
        $saturday = self::saturday();
        $mondayAfter = self::saturday()->addDays(2);

        self::assertTrue(
            BavarianHolidays::addBusinessDays($saturday, 1)
                ->isSameDay($mondayAfter),
            'Skips over sunday'
        );
        self::assertTrue(
            $saturday->isSameDay(self::saturday()),
            'Should not mutate the original date'
        );
    }

    public function testSubMllWorkingDays(): void
    {
        $sunday = self::sunday();
        $thursdayBeforeAllSaints = self::sunday()->subDays(3);
        self::assertTrue(
            BavarianHolidays::subBusinessDays($sunday, 1)
                ->isSameDay($thursdayBeforeAllSaints),
            'Skips over saturday and all saints holiday (01.01.2019)'
        );
        self::assertTrue(
            $sunday->isSameDay(self::sunday()),
            'Should not mutate the original date'
        );
    }

    protected static function businessDayWednesday(): Carbon
    {
        return Carbon::createStrict(2019, 10, 30);
    }

    protected static function karfreitag2019(): Carbon
    {
        return Carbon::createStrict(2019, 4, 19);
    }

    protected static function easterSunday2019(): Carbon
    {
        return Carbon::createStrict(2019, 4, 21);
    }

    protected static function saturday(): Carbon
    {
        return Carbon::createStrict(2019, 11, 2);
    }

    protected static function sunday(): Carbon
    {
        return Carbon::createStrict(2019, 11, 3);
    }

    /**
     * @return iterable<int, array{Carbon}>
     */
    public static function businessDays(): iterable
    {
        yield [self::businessDayWednesday()];
    }

    /**
     * @return iterable<int, array{Carbon}>
     */
    public static function holidays(): iterable
    {
        yield [self::karfreitag2019()];
        yield [self::easterSunday2019()];
    }

    /**
     * @return iterable<int, array{Carbon}>
     */
    public static function weekend(): iterable
    {
        yield [self::saturday()];
        yield [self::sunday()];
    }

    public function testLoadUserDefinedHolidays(): void
    {
        $dayOfTheTentacle = Carbon::createStrict(2019, 8, 22);

        self::assertNull(BavarianHolidays::nameHoliday($dayOfTheTentacle));
        self::assertFalse(BavarianHolidays::isHoliday($dayOfTheTentacle));
        self::assertTrue(BavarianHolidays::isBusinessDay($dayOfTheTentacle));

        $name = 'Day of the Tentacle';
        BavarianHolidays::$loadUserDefinedHolidays = static function (int $year) use ($dayOfTheTentacle, $name): array {
            switch ($year) {
                case 2019:
                    return [BavarianHolidays::dayOfTheYear($dayOfTheTentacle) => $name];
                default:
                    self::fail('Expected the year of the passed in date to be passed');
            }
        };

        self::assertSame($name, BavarianHolidays::nameHoliday($dayOfTheTentacle));
        self::assertTrue(BavarianHolidays::isHoliday($dayOfTheTentacle));
        self::assertFalse(BavarianHolidays::isBusinessDay($dayOfTheTentacle));
    }
}


src/Holidays.php
<?php

declare(strict_types=1);

namespace MLL\Holidays;

use Carbon\Carbon;

class Holidays
{
    public const HOLIDAYS_STATIC = [
        '01.01' => 'Neujahrstag',
        '06.01' => 'Heilige Drei Könige',
        '01.05' => 'Tag der Arbeit',
        '15.08' => 'Maria Himmelfahrt',
        '03.10' => 'Tag der Deutschen Einheit',
        '01.11' => 'Allerheiligen',
        '24.12' => 'Heilig Abend',
        '25.12' => 'Erster Weihnachtstag',
        '26.12' => 'Zweiter Weihnachtstag',
        '31.12' => 'Sylvester',
    ];

    public const SAMSTAG = 'Samstag';
    public const SONNTAG = 'Sonntag';
    public const KARFREITAG = 'Karfreitag';
    public const OSTERSONNTAG = 'Ostersonntag';
    public const OSTERMONTAG = 'Ostermontag';
    public const CHRISTI_HIMMELFAHRT = 'Christi Himmelfahrt';
    public const PFINGSTSONNTAG = 'Pfingstsonntag';
    public const PFINGSTMONTAG = 'Pfingstmontag';
    public const FRONLEICHNAM = 'Fronleichnam';
    public const REFORMATIONSTAG_500_JAHRE_REFORMATION = 'Reformationstag (500 Jahre Reformation)';
    public const BETRIEBSAUSFLUG_2019 = 'Betriebsausflug 2019';

    /**
     * Checks if given date is an MLL working day.
     */
    public static function isMLLWorkingDay(Carbon $date): bool
    {
        return is_null(self::name($date));
    }

    /**
     * Checks if given date is an MLL holiday.
     */
    public static function isMLLHoliday(Carbon $date): bool
    {
        return ! self::isMLLWorkingDay($date);
    }

    /**
     * Returns the name of the holiday if the date happens to land on one.
     */
    public static function name(Carbon $date): ?string
    {
        $holidayMap = self::buildHolidayMap($date);
        $holiday = $holidayMap[self::dayOfTheYear($date)] ?? null;
        if (is_string($holiday)) {
            return $holiday;
        }

        if ($date->isSaturday()) {
            return self::SAMSTAG;
        }

        if ($date->isSunday()) {
            return self::SONNTAG;
        }

        return null;
    }

    public static function addMLLWorkingDays(Carbon $date, int $days): Carbon
    {
        // Make sure we do not mutate the original date
        $copy = $date->clone();

        while ($days > 0) {
            $copy->addDay();
            if (self::isMLLWorkingDay($copy)) {
                $days--;
            }
        }

        return $copy;
    }

    /**
     * Returns a map from day/month to named holidays.
     *
     * @return array<string, string>
     */
    protected static function buildHolidayMap(Carbon $date): array
    {
        $holidays = self::HOLIDAYS_STATIC;

        // dynamic holidays
        $easter = easter_date($date->year);
        $holidays[self::dateFromEaster($easter, -2)] = self::KARFREITAG;
        $holidays[self::dateFromEaster($easter, 0)] = self::OSTERSONNTAG;
        $holidays[self::dateFromEaster($easter, 1)] = self::OSTERMONTAG;
        $holidays[self::dateFromEaster($easter, 39)] = self::CHRISTI_HIMMELFAHRT;
        $holidays[self::dateFromEaster($easter, 49)] = self::PFINGSTSONNTAG;
        $holidays[self::dateFromEaster($easter, 50)] = self::PFINGSTMONTAG;
        $holidays[self::dateFromEaster($easter, 60)] = self::FRONLEICHNAM;

        // exceptional holidays
        if (2017 === $date->year) {
            $holidays['31.10'] = self::REFORMATIONSTAG_500_JAHRE_REFORMATION;
        }

        // Betriebsausflüge
        switch ($date->year) {
            case 2019:
                $holidays['22.07'] = self::BETRIEBSAUSFLUG_2019;
                break;
        }

        return $holidays;
    }

    protected static function dateFromEaster(int $easter, int $daysAway): string
    {
        $date = Carbon::createFromTimestamp($easter)
            ->setTimezone('Europe/Berlin')
            ->addDays($daysAway);

        return self::dayOfTheYear($date);
    }

    protected static function dayOfTheYear(Carbon $date): string
    {
        return $date->format('d.m');
    }
}


src/DateModification.php
<?php

declare(strict_types=1);

namespace MLL\Holidays;

use Carbon\Carbon;

class DateModification
{
    /**
     * @param  callable(Carbon): bool  $shouldCount  Should the given date be added?.
     */
    public static function addDays(Carbon $date, int $days, callable $shouldCount): Carbon
    {
        // Make sure we do not mutate the original date
        $copy = $date->clone();

        while ($days > 0) {
            $copy->addDay();
            if ($shouldCount($copy)) {
                $days--;
            }
        }

        return $copy;
    }

    /**
     * @param  callable(Carbon): bool  $shouldCount  Should the given date be subtracted?.
     */
    public static function subDays(Carbon $date, int $days, callable $shouldCount): Carbon
    {
        // Make sure we do not mutate the original date
        $copy = $date->clone();

        while ($days > 0) {
            $copy->subDay();
            if ($shouldCount($copy)) {
                $days--;
            }
        }

        return $copy;
    }
}


src/BavarianHolidays.php
<?php

declare(strict_types=1);

namespace MLL\Holidays;

use Carbon\Carbon;

/**
 * Some definitions:
 * Holiday: special occasions on a mostly fixed date where there is no work
 * Weekend Day: Saturday and Sunday
 * Business Day: any day that is neither a Holiday nor a Weekend Day.
 */
class BavarianHolidays
{
    public const HOLIDAYS_STATIC = [
        '01.01' => 'Neujahrstag',
        '06.01' => 'Heilige Drei Könige',
        '01.05' => 'Tag der Arbeit',
        '15.08' => 'Maria Himmelfahrt',
        '03.10' => 'Tag der Deutschen Einheit',
        '01.11' => 'Allerheiligen',
        '24.12' => 'Heilig Abend',
        '25.12' => 'Erster Weihnachtstag',
        '26.12' => 'Zweiter Weihnachtstag',
        '31.12' => 'Sylvester',
    ];

    public const SAMSTAG = 'Samstag';
    public const SONNTAG = 'Sonntag';
    public const KARFREITAG = 'Karfreitag';
    public const OSTERSONNTAG = 'Ostersonntag';
    public const OSTERMONTAG = 'Ostermontag';
    public const CHRISTI_HIMMELFAHRT = 'Christi Himmelfahrt';
    public const PFINGSTSONNTAG = 'Pfingstsonntag';
    public const PFINGSTMONTAG = 'Pfingstmontag';
    public const FRONLEICHNAM = 'Fronleichnam';
    public const REFORMATIONSTAG_500_JAHRE_REFORMATION = 'Reformationstag (500 Jahre Reformation)';

    /**
     * Optionally allows users to define extra holidays for a given year.
     *
     * The returned array is expected to be a map from the day of the year
     * (format with @see self::dayOfTheYear()) to holiday names.
     *
     * @example ['23.02' => 'Day of the Tentacle']
     *
     * @var (callable(int): array<string, string>)|null
     */
    public static $loadUserDefinedHolidays;

    /**
     * Checks if given date is a business day.
     */
    public static function isBusinessDay(Carbon $date): bool
    {
        return ! self::isHoliday($date)
            && ! $date->isWeekend();
    }

    /**
     * Checks if given date is a holiday.
     */
    public static function isHoliday(Carbon $date): bool
    {
        return is_string(self::nameHoliday($date));
    }

    /**
     * Returns the name of the holiday if the date happens to land on one.
     * Saturday and Sunday are not evaluated as holiday.
     */
    public static function nameHoliday(Carbon $date): ?string
    {
        $holidayMap = self::buildHolidayMap($date);

        return $holidayMap[self::dayOfTheYear($date)] ?? null;
    }

    /**
     * Returns a new carbon instance with the given number of business days added.
     */
    public static function addBusinessDays(Carbon $date, int $days): Carbon
    {
        return DateModification::addDays($date, $days, [self::class, 'isBusinessDay']);
    }

    /**
     * Returns a new carbon instance with the given number of business days subtracted.
     */
    public static function subBusinessDays(Carbon $date, int $days): Carbon
    {
        return DateModification::subDays($date, $days, [self::class, 'isBusinessDay']);
    }

    /**
     * Returns a map from day/month to named holidays.
     *
     * @return array<string, string>
     */
    protected static function buildHolidayMap(Carbon $date): array
    {
        $holidays = self::HOLIDAYS_STATIC;

        // dynamic holidays
        $easter = easter_date($date->year);
        $holidays[self::dateFromEaster($easter, -2)] = self::KARFREITAG;
        $holidays[self::dateFromEaster($easter, 0)] = self::OSTERSONNTAG;
        $holidays[self::dateFromEaster($easter, 1)] = self::OSTERMONTAG;
        $holidays[self::dateFromEaster($easter, 39)] = self::CHRISTI_HIMMELFAHRT;
        $holidays[self::dateFromEaster($easter, 49)] = self::PFINGSTSONNTAG;
        $holidays[self::dateFromEaster($easter, 50)] = self::PFINGSTMONTAG;
        $holidays[self::dateFromEaster($easter, 60)] = self::FRONLEICHNAM;

        // exceptional holidays
        if (2017 === $date->year) {
            $holidays['31.10'] = self::REFORMATIONSTAG_500_JAHRE_REFORMATION;
        }

        // user-defined holidays
        if (isset(self::$loadUserDefinedHolidays)) {
            $holidays = array_merge(
                $holidays,
                (self::$loadUserDefinedHolidays)($date->year)
            );
        }

        return $holidays;
    }

    protected static function dateFromEaster(int $easter, int $daysAway): string
    {
        $date = Carbon::createFromTimestamp($easter)
            ->setTimezone('Europe/Berlin')
            ->addDays($daysAway);

        return self::dayOfTheYear($date);
    }

    public static function dayOfTheYear(Carbon $date): string
    {
        return $date->format('d.m');
    }
}
